src/rules/expect.ts

```diff
@@ -1,89 +1,73 @@
+import {
+	ExpectRuleContext,
+	MessageIds,
+	Options,
+	defaultOptions,
+	messages,
+} from "../meta.js";
 import { createRule } from "../utils/createRule.js";
 import { isDiagnosticWithStart } from "../utils/diagnostics.js";
 import { lineOfPosition, locForTSNode } from "../utils/locations.js";
 import { updateTypeSnapshot } from "../utils/snapshot.js";
-
-const messages = {
-	ExpectedErrorNotFound: "Expected an error on this line, but found none.",
-	FileIsNotIncludedInTSConfig:
-		'Expected to find a file "{{ fileName }}" present.',
-	Multiple$ExpectTypeAssertions:
-		"This line has 2 or more $ExpectType assertions.",
-	OrphanAssertion: "Can not match a node to this assertion.",
-	SyntaxError: "Syntax Error: {{ message }}",
-	TypeSnapshotDoNotMatch:
-		"Expected type from Snapshot to be: {{ expected }}, got: {{ actual }}",
-	TypeSnapshotNotFound:
-		"Type Snapshot not found. Please consider running ESLint in FIX mode: eslint --fix",
-	TypesDoNotMatch: "Expected type to be: {{ expected }}, got: {{ actual }}",
-};
-
-export type MessageIds = keyof typeof messages;
-
-export interface Options {
-	readonly disableExpectTypeSnapshotFix: boolean;
-}
-
-const defaultOptions: Options = {
-	disableExpectTypeSnapshotFix: false,
-};
-
-type ExpectRuleContext = TSESLint.RuleContext<MessageIds, [Options]>;
+import { resolveVersionsToTest } from "../utils/versions.js";
 
 export const expect = createRule<[Options], MessageIds>({
 	create(context, [options]) {
-		const parserServices = ESLintUtils.getParserServices(context);
-		const { program } = parserServices;
+		// Performance: if the source file doesn't have any triggering comments,
+		// avoid asking for diagnostics or type information altogether.
+		if (!/\$Expect(?:Type|Error|\?)|\^\?/.test(context.sourceCode.text)) {
+			return {};
+		}
+
+		const versionsResolution = resolveVersionsToTest(
+			context,
+			options.versionsToTest ?? [
+				{ name: ts.version, path: require.resolve("typescript") },
+			],
+		);
+		console.log({ versionsResolution });
 
-		const fileName = context.getFilename();
-		const sourceFile = program.getSourceFile(fileName);
-		if (!sourceFile) {
+		if (versionsResolution.error) {
 			context.report({
-				data: {
-					fileName,
-				},
+				...versionsResolution.error,
 				loc: {
-					column: 0,
-					line: 1,
+					end: { column: 0, line: 0 },
+					start: { column: 0, line: 0 },
 				},
-				messageId: "FileIsNotIncludedInTSConfig",
 			});
 			return {};
 		}
 
-		// Performance: if the source file doesn't have any triggering comments,
-		// avoid asking for diagnostics or type information altogether.
-		if (!/\$Expect(?:Type|Error|\?)|\^\?/.test(sourceFile.text)) {
-			return {};
-		}
-
-		const {
-			duplicates,
-			errorLines,
-			syntaxErrors,
-			twoSlashAssertions,
-			typeAssertions,
-		} = parseAssertions(sourceFile);
+		for (const { program, sourceFile } of versionsResolution.versions) {
+			console.log("sourceFile.getText()", sourceFile.getFullText());
```

Comment:
Right now in tests this is logging:

```plaintext
sourceFile.getText() // (stub file for RuleTester programs)
```

🤔 I couldn't figure out how to get the program to read from virtual memory. Grr.

Reply:
Workaround: in 0f142f8 I had it always use the original program if no `versions` are provided. This is probably a nice little performance boost too.